### PR TITLE
Implement Exceptions [14957]

### PR DESCRIPTION
**Pull request by jparisu**
Merge after:
- #1

### CODE REVIEW

File: plugins/datastreamer_plugin/fastdds/Participant.cpp [hunk 1]
```diff
@@ -366,9 +366,19 @@ eprosima::fastrtps::types::DynamicType_ptr Participant::get_type_registered_(
         auto type_object =
                 eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_object(type_name,
                         true);
+        if (!type_object)
+        {
+            IncorrectParamException("Dynamic type not registered");
```
Comment (antonrv) [inline plugins/datastreamer_plugin/fastdds/Participant.cpp]: `throw` missing

File: plugins/datastreamer_plugin/fastdds/Participant.cpp [hunk 2]
```diff
@@ -366,9 +366,19 @@ eprosima::fastrtps::types::DynamicType_ptr Participant::get_type_registered_(
...
+        }
+
         auto type_id =
                 eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_identifier(type_name,
                         true);
+        if (!type_id)
+        {
+            IncorrectParamException("Dynamic type not registered");
```
Comment (antonrv) [inline plugins/datastreamer_plugin/fastdds/Participant.cpp]: `throw` missing

File: plugins/datastreamer_plugin/utils/Exception.hpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    virtual const char* what() const noexcept override
+    {
+        return message_.c_str();
+    }
+
+protected:
+
+    std::string message_;
+};
```
Comment (antonrv) [inline plugins/datastreamer_plugin/utils/Exception.hpp]: Why not inherit the following exceptions from `std::exception`? The additional functionality of `Exception` over `std::exception`, if any, should be documented. If none, `Exception` should not exist, as the cost 60 additional LOCs should be justified.

**Review by antonrv — CHANGES_REQUESTED** [review]
Minor details

File: plugins/datastreamer_plugin/utils/Exception.hpp (hunk 3 above)
Comment (jparisu) [reply plugins/datastreamer_plugin/utils/Exception.hpp]: It adds the functionality to send a custom message, added to class docu

Comment (antonrv) [reply plugins/datastreamer_plugin/utils/Exception.hpp]: that functionality already exists in the standard (good fit for `std::runtime_error`)

So please derive all custom exceptions from it `class UnsupportedException : public std::runtime_error`
or use an alias instead
`using Exception = std::runtime_error;`

**Review by antonrv — APPROVED** [review]

**Pull request merged**